### PR TITLE
Maximum-principle-satisfying WENO scheme

### PR DESCRIPTION
**Pull request by simone-silvestri**
Following https://royalsocietypublishing.org/doi/epdf/10.1098/rspa.2011.0153 (section 5)

At the moment this is just a scheme check, so I have implemented it in 1D (and only in the horizontal direction) 

For the moment this script
```
using Oceananigans

arch = CPU()
grid = RectilinearGrid(arch, size = (100, 100, 1), x = (0, 2π), y = (0, 2π), z=(0, 1), topology = (Periodic, Periodic, Bounded))

free_surface = ImplicitFreeSurface()

model = HydrostaticFreeSurfaceModel(; grid, free_surface,
                                    momentum_advection = WENO5(),
                                    tracer_advection = WENO5(bounds = (0, 1)),
                                    buoyancy = nothing,
                                    tracers = :c)

@inline function linear_u_profile(x, y, z) 
    if x < π/2 || x > 3 * π/2
        return 0.0
    else
        if y > 7/4*π || y < π/4
            return 0.0
        elseif y < 5/4 * π && y > 3/4 * π
            return π/2
        elseif y >= π/4 && y <= 3/4 * π
            return y - π/4
        else
            return π/2 - (y - 5/4 * π)
        end
    end
end

@inline function rectangle_c(x, y, z)
    if x > 2/3 * π && x < 4/3 * π
        if x > 2/3 * π && x < 4/3 * π
            return 1.0
        else
            return 0.0
        end
    else
        return 0.0
    end
end

set!(model, u = linear_u_profile, c = rectangle_c)

Δt = grid.Lx / grid.Nx / maximum(abs, model.velocities.u) * 0.2

for step in 1:10000
    time_step!(model, Δt)
    @info "step $step, time $(model.clock.time)"
end
```
returns 
```
julia> model.tracers.c
100×100×1 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 100×100×1 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 106×106×7 OffsetArray(::Array{Float64, 3}, -2:103, -2:103, -2:4) with eltype Float64 with indices -2:103×-2:103×-2:4
    └── max=1.03919, min=-0.0214752, mean=0.351983
```
for usual `WENO5` 
and 
```
julia> model.tracers.c
100×100×1 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 100×100×1 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 106×106×7 OffsetArray(::Array{Float64, 3}, -2:103, -2:103, -2:4) with eltype Float64 with indices -2:103×-2:103×-2:4
    └── max=0.999906, min=1.09505e-5, mean=0.352073
```
for `WENO5(bounds = (0, 1))`

So it seems like this scheme, even in its 1D formulation, might  help with our tracer bound issues.

### CODE REVIEW

**Comment by simone-silvestri** [comment]
This is a heatmap of the tracer after 10000 time steps. The top is `WENO5`, the bottom `WENO5(positivity_preserving = true)`. The out-of-bounds cells are substituted with a NaN

positivity_preserving

**Comment by simone-silvestri** [comment]
Same plot obtained with a second order centered scheme for tracers

positivity_preserving-2

**Comment by francispoulin** [comment]
This is great to see!

I imagine we could use this for the height field in `ShallowWaterModel` to allow for wetting and drying for a single layer.  And also to prevent layer collapse in the case of multiple layers.

**Comment by simone-silvestri** [comment]
yeah, I guess in that case you don't want to set an upper bound (`bounds = (0, Inf)`)

**Comment by glwagner** [comment]
Just a small plea to make this implementation compatible with / moves us towards the refactor envisioned in #2454

File: src/Advection/positivity_preserving_tracer_advection_operators.jl [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+@inline function div_Uc(i, j, k, grid, advection::BoundPreservingScheme, U, c)
+
+    div_x = bounded_tracer_flux_divergence_x(i, j, k, grid, advection, U.u, c)
+    div_y = bounded_tracer_flux_divergence_y(i, j, k, grid, advection, U.v, c)
+    div_z = bounded_tracer_flux_divergence_z(i, j, k, grid, advection, U.w, c)
+
+    return 1/Vᶜᶜᶜ(i, j, k, grid) * (div_x + div_y + div_z)
+end
+
+@inline function bounded_tracer_flux_divergence_x(i, j, k, grid, advection::BoundPreservingScheme, u, c)
```
Comment (francispoulin) [inline src/Advection/positivity_preserving_tracer_advection_operators.jl:19]: This looks great!  It seems like these three funcionts are almost identical except for u -> v -> w.  Is that true?

If yes, I wonder if there is away to merge these to one, rather than having three functions that do the same thing?

File: src/Advection/weno_fifth_order.jl [hunk 2]
```diff
@@ -26,7 +30,7 @@ Weighted Essentially Non-Oscillatory (WENO) fifth-order advection scheme.
 
 $(TYPEDFIELDS)
 """
-struct WENO5{FT, XT, YT, ZT, XS, YS, ZS, VI, WF} <: AbstractUpwindBiasedAdvectionScheme{2}
+struct WENO5{FT, XT, YT, ZT, XS, YS, ZS, VI, WF, PP} <: AbstractUpwindBiasedAdvectionScheme{2}
```
Comment (glwagner) [inline src/Advection/weno_fifth_order.jl:33]: Is this the design we will want when we do #2454 ? If we can avoid digging ourselves into a deeper hole we should, but maybe it's not possible without a lot of effort...

**Review by glwagner — APPROVED** [review]

**Comment by glwagner** [comment]
To support different bounds for different tracers, let's allow the advection scheme to be a `NamedTuple`. Then pass the advection scheme specific to each tracer into to the tracer tendency kernel.

**Comment by simone-silvestri** [comment]
I think it is already like this, we can specify
```
model = HydrostaticFreeSurfaceModel(tracer_advection = (c = WENO5(bounds = (0, 1), T = WENO5()), tracers = (:c, :T))
```
I guess the nonhydrostatic model does not do the same

**Comment by simone-silvestri** [comment]
I guess it is better to change the `ShallowWaterModel` in #2522 as a lot is changing already

File: src/Advection/weno_fifth_order.jl (hunk 2 above)
Comment (simone-silvestri) [reply src/Advection/weno_fifth_order.jl:33]: I am not still completely sure how #2454 will go along, this change only modifies the `div_Uc` so I guess it might be easy to adapt in the future

**Comment by simone-silvestri** [comment]
I like it! I changed the NonhydrostaticModel to be the same

**Comment by glwagner** [comment]
> I like it! I changed the NonhydrostaticModel to be the same.
> 
> A way to facilitate users is that we leave the option to specify a global `advection` (so that we keep the keyword) or two separate `momentum_advection` and `tracer_advection`. I would rather just change it to be consistent with the HydrostaticFreeSurfaceModel directly.
> 
> what do you guys think?

I think we should do that in a different PR because it's a major, major breaking change to the most popular model!

File: src/Models/NonhydrostaticModels/nonhydrostatic_model.jl [hunk 3]
```diff
@@ -50,7 +55,8 @@ end
 """
     NonhydrostaticModel(;     grid,
                               clock = Clock{eltype(grid)}(0, 0, 1),
-                          advection = CenteredSecondOrder(),
+                 momentum_advection = CenteredSecondOrder(),
+                   tracer_advection = CenteredSecondOrder(),
```
Comment (glwagner) [inline src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:59]: I prefer advection, and we should do this in a different PR.

**Review by glwagner — CHANGES_REQUESTED** [review]

File: src/Models/NonhydrostaticModels/nonhydrostatic_model.jl (hunk 3 above)
Comment (simone-silvestri) [reply src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:59]: Ok, then we can change it in a different PR (for the moment then separate `bounds` can be used only in the `HydrostaticFreeSurfaceModel`)

File: src/Advection/positivity_preserving_tracer_advection_operators.jl [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    c₊ᴸ =  _left_biased_interpolate_xᶠᵃᵃ(i+1, j, k, grid, advection, c)
+    c₊ᴿ = _right_biased_interpolate_xᶠᵃᵃ(i+1, j, k, grid, advection, c)
+    c₋ᴸ =  _left_biased_interpolate_xᶠᵃᵃ(i,   j, k, grid, advection, c)
+    c₋ᴿ = _right_biased_interpolate_xᶠᵃᵃ(i,   j, k, grid, advection, c)
+
+    p̃   =  (cᵢⱼ - ω̂₁ * c₋ᴿ - ω̂ₙ * c₊ᴸ) / (1 - 2ω̂₁)
+    M   = max(p̃, c₊ᴸ, c₋ᴿ) 
+    m   = min(p̃, c₊ᴸ, c₋ᴿ) 
+    θ   = min(abs((upper_limit - cᵢⱼ)/(M - cᵢⱼ + ε₂)), abs((lower_limit - cᵢⱼ)/(m - cᵢⱼ + ε₂)), 1.0)
```
Comment (glwagner) [inline src/Advection/positivity_preserving_tracer_advection_operators.jl:40]:
```suggestion
    θ   = min(abs((upper_limit - cᵢⱼ)/(M - cᵢⱼ + ε₂)), abs((lower_limit - cᵢⱼ)/(m - cᵢⱼ + ε₂)), one(grid))
```

File: src/Advection/positivity_preserving_tracer_advection_operators.jl [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    c₊ᴸ =  _left_biased_interpolate_yᵃᶠᵃ(i, j+1, k, grid, advection, c)
+    c₊ᴿ = _right_biased_interpolate_yᵃᶠᵃ(i, j+1, k, grid, advection, c)
+    c₋ᴸ =  _left_biased_interpolate_yᵃᶠᵃ(i, j,   k, grid, advection, c)
+    c₋ᴿ = _right_biased_interpolate_yᵃᶠᵃ(i, j,   k, grid, advection, c)
+
+    p̃   =  (cᵢⱼ - ω̂₁ * c₋ᴿ - ω̂ₙ * c₊ᴸ) / (1 - 2ω̂₁)
+    M   = max(p̃, c₊ᴸ, c₋ᴿ) 
+    m   = min(p̃, c₊ᴸ, c₋ᴿ) 
+    θ   = min(abs((upper_limit - cᵢⱼ)/(M - cᵢⱼ + ε₂)), abs((lower_limit - cᵢⱼ)/(m - cᵢⱼ + ε₂)), 1.0)
```
Comment (glwagner) [inline src/Advection/positivity_preserving_tracer_advection_operators.jl:64]:
```suggestion
    θ   = min(abs((upper_limit - cᵢⱼ)/(M - cᵢⱼ + ε₂)), abs((lower_limit - cᵢⱼ)/(m - cᵢⱼ + ε₂)), one(grid))
```

File: src/Advection/positivity_preserving_tracer_advection_operators.jl [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    c₊ᴸ =  _left_biased_interpolate_zᵃᵃᶠ(i, j, k+1, grid, advection, c)
+    c₊ᴿ = _right_biased_interpolate_zᵃᵃᶠ(i, j, k+1, grid, advection, c)
+    c₋ᴸ =  _left_biased_interpolate_zᵃᵃᶠ(i, j, k,   grid, advection, c)
+    c₋ᴿ = _right_biased_interpolate_zᵃᵃᶠ(i, j, k,   grid, advection, c)
+
+    p̃   =  (cᵢⱼ - ω̂₁ * c₋ᴿ - ω̂ₙ * c₊ᴸ) / (1 - 2ω̂₁)
+    M   = max(p̃, c₊ᴸ, c₋ᴿ) 
+    m   = min(p̃, c₊ᴸ, c₋ᴿ) 
+    θ   = min(abs((upper_limit - cᵢⱼ)/(M - cᵢⱼ + ε₂)), abs((lower_limit - cᵢⱼ)/(m - cᵢⱼ + ε₂)), 1.0)
```
Comment (glwagner) [inline src/Advection/positivity_preserving_tracer_advection_operators.jl:88]:
```suggestion
    θ   = min(abs((upper_limit - cᵢⱼ)/(M - cᵢⱼ + ε₂)), abs((lower_limit - cᵢⱼ)/(m - cᵢⱼ + ε₂)), one(grid))
```

File: src/Advection/weno_fifth_order.jl (hunk 2 above)
Comment (glwagner) [reply src/Advection/weno_fifth_order.jl:33]: In #2454 we will disentangle vector invariant from WENO5, so `VI` will go away. It might still make sense to blend `WENO5` with flux-limited WENO5, however. It's just that if we have nth order WENO, we would also have to figure out how to do nth order flux limited WENO if they are entangled.

**Review by glwagner — APPROVED** [review]

File: src/Advection/weno_fifth_order.jl (hunk 2 above)
Comment (simone-silvestri) [reply src/Advection/weno_fifth_order.jl:33]: I think the limiter will be independent on the order of the original scheme. You always want to blend your scheme into a first order upwind (when crossing bounds)

Comment (glwagner) [reply src/Advection/weno_fifth_order.jl:33]: ok, great!
